### DISCUSSION TITLE
ReactiveUI/ItemsControl  list performance

### DISCUSSION THREAD

**Discussion by ctrl-alt-d** [AvaloniaUI/Avalonia#6417] [Q&A] [answered]
My app has a list of about 200 items. When I refresh item list, it takes about 2 seconds to refresh UI. Is there a way to show a spinner or improve performance?

![](https://i.stack.imgur.com/sqIKs.gif)

**What I have tried**

I tried to bind UI to a `ReadOnlyObservableCollection` cache pattern, but this doesn't solve performance issues. It is not related with backend operations.

```c#

        public CentreSetViewModel(bool? modeLookup = null)
        {
            SourceItems
                .ToObservableChangeSet(t=>t.Id)
                .Bind(out _MyItems)
                .Subscribe();

            // observe toggle button
        }

        public ObservableCollectionExtended<CentreRowViewModel> SourceItems { get; } = new();
        private readonly ReadOnlyObservableCollection<CentreRowViewModel> _MyItems;
        public ReadOnlyObservableCollection<CentreRowViewModel> MyItems => _MyItems;
```

**More info**

The toggle button is bound to this:

```c#

            // observe toggle button
            this
                .WhenAnyValue(x => x.NomesActius)
                .ObserveOn(RxApp.MainThreadScheduler)
                .Subscribe(nomesActius => LoadCentres(nomesActius))
                ;


            // ...

        protected virtual async void LoadCentres(bool nomesActius)
        {
            SourceItems.Clear();
            var newItems =
                ( await backendCall() )
                .Select(x => new CentreRowViewModel(x, ModeLookup));
            SourceItems.AddRange(newItems);
        }

```

To reproduce:

* Clone my app: `git clone https://github.com/ctrl-alt-d/eaprecull.git`
* `cd eaprecull`
* `git checkout trycoll`
* Create some demo data: `dotnet run --project CreateDemoData`
* Start app: `dotnet run --project UI.ER.AvaloniaUI`
* Click on right menu, click `Centres`, click toggle button.

Thanks!

( I also posted the question on SO https://stackoverflow.com/questions/68781907/reactiveui-refresh-list-performance )

**Comment by Gillibald** [comment, marked as answer]
Use a ListBox instead that supports virtualization

**Reply by ctrl-alt-d** [reply]
Hi! Impressive performance with ListBox. Thanks, @Gillibald !!!

(I open new discussion with issues related with virtualization: https://github.com/AvaloniaUI/Avalonia/discussions/6420 )
